src/allmydata/web/root.py

```diff
@@ -29,30 +39,40 @@
     create_private_tree,
 )
 
-class URIHandler(RenderMixin, rend.Page):
-    # I live at /uri . There are several operations defined on /uri itself,
-    # mostly involved with creation of unlinked files and directories.
+class URIHandler(resource.Resource, object):
+    """
+    I live at /uri . There are several operations defined on /uri itself,
+    mostly involved with creation of unlinked files and directories.
+    """
 
     def __init__(self, client):
-        rend.Page.__init__(self, client)
+        super(URIHandler, self).__init__()
         self.client = client
 
-    def render_GET(self, ctx):
-        req = IRequest(ctx)
+    def render_GET(self, req):
+        """
+        Historically, accessing this via "GET /uri?uri=<capabilitiy>"
+        was/is a feature -- which simply redirects to the more-common
+        "GET /uri/<capability>". New code should use /uri/<cap>
+        """
         uri = get_arg(req, "uri", None)
         if uri is None:
             raise WebError("GET /uri requires uri=")
-        there = url.URL.fromContext(ctx)
-        there = there.clear("uri")
-        # I thought about escaping the childcap that we attach to the URL
-        # here, but it seems that nevow does that for us.
-        there = there.child(uri)
-        return there
-
-    def render_PUT(self, ctx):
-        req = IRequest(ctx)
-        # either "PUT /uri" to create an unlinked file, or
-        # "PUT /uri?t=mkdir" to create an unlinked directory
+        base = URLPath.fromRequest(req)
+        args = '&'.join([
+            '{}={}'.format(name, vals[0])
+            for name, vals in req.args.items()
+            if name != 'uri'
+        ])
+        if args:
+            args = '?{}'.format(args)
+        return redirectTo(b'{}{}'.format(base.child(urllib.quote(uri)), args), req)
```

Comment:
This all seems crazy .. I think what Nevow was doing was essentially just rendering the right child "inline", but my stabs at that cause confusion because I can't seem to remove the `?uri=` query-arg.